plugin/evm/vm_test.go

```diff
@@ -2110,3 +2113,157 @@ func TestCreateHandlers(t *testing.T) {
 		require.ErrorIs(t, elem.Error, rpc.ErrMissingBatchResponse)
 	}
 }
+
+// deployContract deploys the provided EVM bytecode using a prefunded test account
+// and returns the created contract address. It is reusable for any contract code.
+func deployContract(t *testing.T, vm *VM, gasPrice *big.Int, code []byte) common.Address {
+	callerAddr := vmtest.TestEthAddrs[0]
+	callerKey := vmtest.TestKeys[0]
+
+	nonce := vm.txPool.Nonce(callerAddr)
+	tx := types.NewContractCreation(nonce, big.NewInt(0), 1000000, gasPrice, code)
+	signedTx, err := types.SignTx(tx, types.NewEIP155Signer(vm.chainConfig.ChainID), callerKey.ToECDSA())
+	require.NoError(t, err)
+
+	for _, err := range vm.txPool.AddRemotesSync([]*types.Transaction{signedTx}) {
+		require.NoError(t, err)
+	}
+
+	blk, err := vm.BuildBlock(context.Background())
+	require.NoError(t, err)
+	require.NoError(t, blk.Verify(context.Background()))
+	require.NoError(t, vm.SetPreference(context.Background(), blk.ID()))
+	require.NoError(t, blk.Accept(context.Background()))
+
+	ethBlock := blk.(*chain.BlockWrapper).Block.(*wrappedBlock).ethBlock
+	receipts := vm.blockChain.GetReceiptsByHash(ethBlock.Hash())
+	require.Len(t, receipts, len(ethBlock.Transactions()))
+
+	found := false
+	for i, btx := range ethBlock.Transactions() {
+		if btx.Hash() == signedTx.Hash() {
+			found = true
+			require.Equal(t, uint64(1), receipts[i].Status) // 1 = success
+			break
+		}
+	}
+	require.True(t, found)
+
+	return crypto.CreateAddress(callerAddr, nonce)
+}
+
+func TestDelegatePrecompile_BehaviorAcrossUpgrades(t *testing.T) {
+	// Mirror params.invalidateDelegateUnix here - if exported and
+	// imported we get a circular dependency
+	const invalidateDelegateUnix = 1754107200
+
+	tests := []struct {
+		name                  string
+		fork                  upgradetest.Fork
+		deployGasPrice        *big.Int
+		txGasPrice            *big.Int
+		preDeploySetTime      func(vm *VM)
+		setTime               func(vm *VM)
+		refillCapacityFortuna bool
+		wantIncluded          bool
+		wantReceiptStatus     uint64
+	}{
+		{
+			name:           "granite_should_revert",
+			fork:           upgradetest.Granite,
+			deployGasPrice: vmtest.InitialBaseFee,
+			txGasPrice:     vmtest.InitialBaseFee,
+			setTime: func(vm *VM) {
+				vm.clock.Set(time.Unix(2000000000, 0))
+			},
+			refillCapacityFortuna: false,
+			wantIncluded:          true,
+			wantReceiptStatus:     0,
+		},
+		{
+			name:           "fortuna_post_cutoff_should_invalidate",
+			fork:           upgradetest.Fortuna,
+			deployGasPrice: big.NewInt(ap0.MinGasPrice),
+			txGasPrice:     big.NewInt(ap0.MinGasPrice),
+			setTime: func(vm *VM) {
+				// Ensure timestamp is >= cutoff by adding the necessary delta
+				now := vm.clock.Time().Unix()
+				if now < int64(invalidateDelegateUnix+1) {
+					delta := (int64(invalidateDelegateUnix+1) - now)
+					vm.clock.Set(vm.clock.Time().Add(time.Duration(delta) * time.Second))
+				}
```

Comment:
```suggestion
				if now := vm.clock.Time().Unix(); now <= invalidateDelegateUnix {
					delta := time.Duration(invalidateDelegateUnix - now + 1) * time.Second
					vm.clock.Set(vm.clock.Time().Add(delta))
				}
```

Rationale:
* The `int64` casting was redundant
* `x < y + 1` is clearer as `x <= y`
* Scope variables (`now`) to signal to readers how long they'll need to remember them